src/Binding.lua

```diff
@@ -146,4 +159,33 @@ function Binding.create(initialValue)
 	return binding, setter
 end
 
+--[[
+	Creates a new binding which updates when any of the upstream bindings
+	updates, which can be further mapped into any value. This function will
+	be exposed to users of Roact.
+]]
+function Binding.join(bindings)
+	local joinedBinding, setter = Binding.create(mapBindingsToValues(bindings))
+	local internalData = joinedBinding[InternalData]
+
+	local upstreamConnections = {}
+	internalData.upstreamConnections = upstreamConnections
```

Comment:
Can we change `upstreamBinding` to work through this same mechanism (a table instead of a single entry) to consolidate the disconnection logic?

Some things that would fall out of this:
1. `upstreamDisconnect` can become static instead of being a closure we generate in a couple places
2. `Binding:map` changes to fill `upstreamConnections` instead of setting `upstreamBinding`